neat_pvd.c

```diff
@@ -2,17 +2,199 @@
+    ldns_rdf *dns_record = NULL;
+
+    struct pvd* pvd = (struct pvd *) malloc(sizeof(struct pvd));
+    LIST_INIT(&pvd_infos);
+
+    for (int i = 0; i < nb_txt; i++) {
+        rr = ldns_rr_list_rr(pvd_txt_list, i);
+        dns_record = ldns_rr_set_rdf(rr, NULL, 0);
+        txt_record = strdup(ldns_rdf2str(dns_record));
+
+        // Removing quotes if any
+        if (txt_record[0] == '"' && txt_record[strlen(txt_record)-1] == '"') {
+            txt_record[strlen(txt_record)-1] = 0;
+            txt_record++;
+        }
+
+        pvd_info = (struct pvd_info *) malloc(sizeof(struct pvd_info));
+        pvd_info->key = strsep(&txt_record, "=");
+        pvd_info->value = txt_record;
+
+        LIST_INSERT_HEAD(&(pvd_infos), pvd_info, next_info);
+    }
+
+    if (nb_txt > 0) {
+        pvd->infos = pvd_infos;
+        LIST_INSERT_HEAD(&(src_addr->pvds), pvd, next_pvd);
+    }
+
+    ldns_rr_list_deep_free(pvd_txt_list);
+}
 
 static void neat_pvd_handle_newaddr(struct neat_ctx *nc,
-                                         void *p_ptr,
-                                         void *data)
+                                    void *p_ptr,
+                                    void *data)
 {
+    if (LIST_EMPTY(&(nc->resolver->server_list))) {
+        // No DNS servers
+        return;
+    }
+
+    struct in_addr dns_addr4;
+    struct in6_addr dns_addr6;
+    struct neat_addr *src_addr = (struct neat_addr*) data;
+    struct neat_resolver_server *dns_server;
+    int i, nb_ptr;
+    ldns_resolver *resolver;
+    ldns_rdf *domain;
+    ldns_rdf *dns_src;
+    ldns_rdf *ptr;
+    ldns_pkt *p;
+    ldns_rr_list *pvd_txt_list;
+    ldns_rr_list *pvd_ptr_list;
+    ldns_rr *rr;
+    ldns_rdf *dns_record = NULL;
+    char* ptr_record;
+    char* reverse_ip = compute_reverse_ip(src_addr);
+
+    if (strlen(reverse_ip) == 0) {
+        return;
+    }
+
+    LIST_INIT(&(src_addr->pvds));
 
+    LIST_FOREACH(dns_server, &(nc->resolver->server_list), next_server) {
+        // Avoid static servers
+        if (dns_server->mark != NEAT_RESOLVER_SERVER_ACTIVE) {
+            continue;
+        }
+
+        struct sockaddr_storage *dns_addr = &(dns_server->server_addr);
+        if (dns_addr->ss_family == AF_INET6) {
+            dns_addr6 = ((struct sockaddr_in6*) dns_addr)->sin6_addr;
+            dns_src = ldns_rdf_new_frm_data(LDNS_RDF_TYPE_AAAA, 16, &dns_addr6);
+        } else if (dns_addr->ss_family == AF_INET) {
+            dns_addr4 = ((struct sockaddr_in*) dns_addr)->sin_addr;
+            dns_src = ldns_rdf_new_frm_data(LDNS_RDF_TYPE_A, 4, &dns_addr4);
+        } else {
+            continue;
+        }
+
+        domain = ldns_dname_new_frm_str(reverse_ip);
+        resolver = ldns_resolver_new();
+        ldns_resolver_push_nameserver(resolver, dns_src);
+
+        // Performing DNS query to 'dns_addr', for PTR records of 'reverse_ip'
+        p = ldns_resolver_query(resolver, domain, LDNS_RR_TYPE_PTR, LDNS_RR_CLASS_IN, LDNS_RD);
```

Comment:
This is a blocking request. neat_resolver.c contains code showing how to do async DNS queries. Basically what you need to do is to create a handle + pointer to data structure for each request, store the pointer in the handle and then wait for the callback to be called and parse the result.